File: packages/python/examples/algorithms_demo.py
"""
Example demonstrating all three rate limiting algorithms.
"""

from halt import RateLimiter, InMemoryStore, Policy, Algorithm, KeyStrategy
import time


def test_algorithm(algorithm: Algorithm, name: str, sliding_precision: int | None = None):
    """Test a specific algorithm."""
    print(f"\n{'='*60}")
    print(f"Testing {name}")
    print('='*60)

    policy_kwargs = dict(
        name=name.lower().replace(' ', '_'),
        limit=5,
        window=10,  # 10 seconds
        algorithm=algorithm,
        key_strategy=KeyStrategy.IP,
    )

    if algorithm == Algorithm.SLIDING_WINDOW and sliding_precision is not None:
        policy_kwargs["sliding_precision"] = sliding_precision
        print(f"Sliding precision: {sliding_precision}")

    policy = Policy(**policy_kwargs)
    
    limiter = RateLimiter(store=InMemoryStore(), policy=policy)
    
    # Mock request
    class MockRequest:
        def __init__(self):
            self.client = type('obj', (object,), {'host': '192.168.1.100'})
    
    request = MockRequest()
    
    # Make 7 requests
    for i in range(7):
        decision = limiter.check(request)
        
        status = "✅ ALLOWED" if decision.allowed else "❌ BLOCKED"
        print(f"Request {i+1}: {status}")
        print(f"  Limit: {decision.limit}")
        print(f"  Remaining: {decision.remaining}")
        print(f"  Reset at: {decision.reset_at}")
        
        if not decision.allowed and decision.retry_after:
            print(f"  Retry after: {decision.retry_after}s")
        
        print()
        time.sleep(0.5)  # Small delay between requests


if __name__ == "__main__":
    print("Halt Rate Limiting - Algorithm Comparison")
    print("This demo shows how different algorithms behave")
    print("Policy: 5 requests per 10 seconds")
    
    # Test Token Bucket
    test_algorithm(Algorithm.TOKEN_BUCKET, "Token Bucket")
    
    # Test Fixed Window
    test_algorithm(Algorithm.FIXED_WINDOW, "Fixed Window")
    
    # Test Sliding Window with custom precision
    test_algorithm(Algorithm.SLIDING_WINDOW, "Sliding Window", sliding_precision=20)
    
    print("\n" + "="*60)
    print("Algorithm Comparison Complete!")
    print("="*60)
    print("\nKey Differences:")
    print("- Token Bucket: Smooth refill, handles bursts well")
    print("- Fixed Window: Simple, but can allow 2x at boundaries")
    print("- Sliding Window: Most accurate, higher memory usage")

File: packages/typescript/examples/algorithms-demo.ts
/**
 * Example demonstrating all three rate limiting algorithms.
 */

import { RateLimiter, InMemoryStore, Policy, Algorithm, KeyStrategy } from '../src';

function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function testAlgorithm(
    algorithm: Algorithm,
    name: string,
    slidingPrecision?: number
): Promise<void> {
    console.log('\n' + '='.repeat(60));
    console.log(`Testing ${name}`);
    console.log('='.repeat(60));

    const policy: Policy = {
        name: name.toLowerCase().replace(/\s+/g, '_'),
        limit: 5,
        window: 10, // 10 seconds
        algorithm,
        keyStrategy: KeyStrategy.IP,
        ...(algorithm === Algorithm.SLIDING_WINDOW && slidingPrecision
            ? { slidingPrecision }
            : {}),
    };

    if (algorithm === Algorithm.SLIDING_WINDOW && slidingPrecision) {
        console.log(`Sliding precision: ${slidingPrecision}`);
    }

    const limiter = new RateLimiter({
        store: new InMemoryStore(),
        policy,
    });

    // Mock request
    const request = {
        socket: { remoteAddress: '192.168.1.100' },
        headers: {},
    };

    // Make 7 requests
    for (let i = 0; i < 7; i++) {
        const decision = await limiter.check(request);

        const status = decision.allowed ? '✅ ALLOWED' : '❌ BLOCKED';
        console.log(`Request ${i + 1}: ${status}`);
        console.log(`  Limit: ${decision.limit}`);
        console.log(`  Remaining: ${decision.remaining}`);
        console.log(`  Reset at: ${decision.resetAt}`);

        if (!decision.allowed && decision.retryAfter) {
            console.log(`  Retry after: ${decision.retryAfter}s`);
        }

        console.log();
        await sleep(500); // Small delay between requests
    }
}

async function main() {
    console.log('Halt Rate Limiting - Algorithm Comparison');
    console.log('This demo shows how different algorithms behave');
    console.log('Policy: 5 requests per 10 seconds');

    // Test Token Bucket
    await testAlgorithm(Algorithm.TOKEN_BUCKET, 'Token Bucket');

    // Test Fixed Window
    await testAlgorithm(Algorithm.FIXED_WINDOW, 'Fixed Window');

    // Test Sliding Window with custom precision
    await testAlgorithm(Algorithm.SLIDING_WINDOW, 'Sliding Window', 20);

    console.log('\n' + '='.repeat(60));
    console.log('Algorithm Comparison Complete!');
    console.log('='.repeat(60));
    console.log('\nKey Differences:');
    console.log('- Token Bucket: Smooth refill, handles bursts well');
    console.log('- Fixed Window: Simple, but can allow 2x at boundaries');
    console.log('- Sliding Window: Most accurate, higher memory usage');
}

main().catch(console.error);
